method/Spatial-mgcn/Spatial-mgcn.py

```diff
@@ -0,0 +1,254 @@
+    required=False,
+)
+
+# Uncomment to calculate ARI every epoch as in original implementation
+if CALC_ARI:
+    parser.add_argument(
+        "-g", "--groundtruth",
+        help="Groundtruth.",
+        required=False,
+    )
+
+args = parser.parse_args()
+
+from pathlib import Path
+
+out_dir = Path(args.out_dir)
+
+# Output files
+label_file = out_dir / "domains.tsv"
+embedding_file = out_dir / "embedding.tsv"
+
+n_clusters = args.n_clusters
+technology = args.technology
+seed = args.seed
+
+## Your code goes here
+import json
+
+with open(args.config, "r") as f:
+    config = json.load(f)
+
+import random
+import numpy as np
+import pandas as pd
+import torch
+import tempfile
+import os
+import sys
+import scanpy as sc
+import scipy as sp
+import torch.optim
+import sklearn.cluster
+import sklearn.metrics
+import tqdm
+
+def get_anndata(args):
+    import anndata as ad
+    
+    from PIL import Image
+    import scipy.io
+
+    X = sp.io.mmread(args.matrix)
+    if sp.sparse.issparse(X):
+        X = X.tocsr()
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    # Filter
+    if "selected" in observations.columns:
+        X = X[observations["selected"].to_numpy().nonzero()[0], :]
+        observations = observations.loc[lambda df: df["selected"]]
+    if "selected" in features.columns:
+        X = X[:, features["selected"].to_numpy().nonzero()[0]]
+        features = features.loc[lambda df: df["selected"]]
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData( # Rename spatial_pixel to spatial for Spatial-mgcn
+        X=X, obs=observations, var=features, obsm={"spatial": coordinates}
+    )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.image))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+
+adata = get_anndata(args)
+
+if CALC_ARI:
+    labels = pd.read_table(args.groundtruth, index_col=0)
+    adata = adata[adata.obs_names.isin(labels.index)]
+    labels = labels.loc[adata.obs_names]
+
+# Set seed
+random.seed(seed)
+torch.manual_seed(seed)
+np.random.seed(seed)
+torch.cuda.manual_seed(seed)
+
+# Work in a temprary folder
+with tempfile.TemporaryDirectory() as tmpdir:
+    gitdir = f"{str(tmpdir)}/Spatial-MGCN"
+
+    # Clone the repository to the specific commit
+    os.system(
+    f"""
+    git clone https://github.com/cs-wangbo/Spatial-MGCN.git {gitdir}
+    cd {gitdir} 
+    git reset --hard cf4412d
+    """)
+
+    # Set working directory as Spatial-MGCN directory
+    sys.path.append(f"{gitdir}/Spatial-MGCN")
+    import utils
+    import models
+
+    # Normalize data: min_cells, calculate HVG and scale
+    sc.pp.filter_genes(adata, min_cells=100)
+    sc.pp.highly_variable_genes(adata, flavor="seurat_v3", n_top_genes=config["fdim"])
```

Comment:
Is there a reason not to use existing feature selection?

Reply:
I changed it to the existing feature selection now. But I left their feature selection as comments to make it easier to compare the performance of both ways.